### ISSUE TITLE
MUC bookmark support?

### ISSUE THREAD

**Issue by poVoq** [cmeng-git/atalk-android#39]
Many XMPP clients supporrt server side bookmarking of MUCs, so that you can easily join MUCs added on another client or just saving them for later without actually joining them.

Would be great if this could also be supported in aTalk.

**Comment by cmeng-git** [comment]
I see that there are two XEP standards describing bookmarks.
1. XEP-0048: Bookmarks and
2. XEP-0402: Bookmarks 2 (This Time it's Serious)

Both standards describes the purpose of the bookmarks. But I am not sure how exactly is one uses the bookmarks? Will you be able to provide a description on how you use it.

I have conversations installed on my device, and see there is a long list of muc rooms listed under the tab BOOKMARKS. I do not know how all these muc rooms get listed under this BOOKMARkS, for which I do take any prior actions on any of them. Also most of these muc rooms are already being destroyed and not in use. The BOOKMARKS really confuses me.

**Comment by poVoq** [comment]
Some clients allow setting bookmarks specifically, but I think in conversations these are saved for MUCs that you have previously joined and then left again. It's quite covenient if you want to remember a MUC that you don't want to follow all the time and also for cross client MUC address sharing.

**Comment by weiss** [comment]
>  I do not know how all these muc rooms get listed under this BOOKMARkS

This basically must have been you adding them with some client long time ago or something.  In theory, the admin could've done this of course, but this doesn't seem very likely.

> I see that there are two XEP standards describing bookmarks.

I'm not aware of any client implementing XEP-0402, and there's some consensus in at least my corner of the community that it does more harm than good.  The problem is there's a _third_ way of doing it (welcome to XMPP!): XEP-0223 (i.e., PEP).  Traditionally, most clients used XEP-0048.  The modern solution is XEP-0223, which is used at least by [Converse.js][1] and supported by [Conversations][2].  On the plus side, servers can transparently sync XEP-0048 and XEP-0223.  These days, ejabberd does this out of the box, and there's a module for Prosody to do this.  Unfortunately, this can't be done for XEP-0402.

Long story short: I'd recommend implementing XEP-0223 and relying on server support for compatibilty with XEP-0048.

> Will you be able to provide a description on how you use it.

I think a modern client should automatically check for bookmarks and join them (if the `autojoin` flag is set on the bookmark).  If the user is invited to a new room (or manually joins one), the client should auto-bookmark it.  Apart from that, you could optionally expose the list of bookmarks to the user just like you expose the roster.

**Comment by cmeng-git** [comment]
I have checked and found that Gajim has a more user friendly UI in its implementation.
I will take a look on how to implement this based on Smack library.

**Comment by weiss** [comment]
> Gajim has a more user friendly UI in its implementation.

Gajim's UI is surely nice for power users (i.e., Gajim's target audience) who might want to view/edit their bookmarks manually, but I don't think typical mobile users want/need this.  Personally I'd just use the bookmarks mechanism to sync the list of joined rooms server-side and maybe expose the list of known groups to the known user; i.e., as a "roster for groups" (similar to how Conversations does it).  Anything beyond that seems too complex to me, UX-wise.

**Comment by cmeng-git** [comment]
Actually, I have already half way through implementation.

![image](https://user-images.githubusercontent.com/7959606/59340558-ce47ae80-8d38-11e9-93d5-e033f2c46550.png)

**Comment by weiss** [comment]
That was quick :+1:

> Need to give a thought whether to abandon it and start over again.

Oh well, power users will like it, don't give too much on my 2 Cent above.  You can always wait for more user feedback I guess.

I think the important bits are to (1) actually perform the auto-join and to (2) automatically create a new bookmark (with `autojoin=true`) whenever the user is invited to a new room (or joins one manually).

**Comment by cmeng-git** [comment]
Conversations Bookmarks context menu have two options i.e.
1. Delete boorkmark
2. Share URI with ....

Need your assistance on the following:
1. Delete of bookmark only remove server bookmark instead of setting Auto Join to false.
a. How is user join the chatroom again, since there is no reference to it anymore after the bookmark is deleted.
b. No sure what happen to the all history messages in the database. Should it be removed also?
The user will no see the history messages again unless he is being invited to join the chatroom again.

2. Share URI with ....
I see it is only link e.g. xmpp:chatroom@conference.atalk.org?join
a. How is the link info being used by the recipient?
b. Conversations also does not allow click to join the chatRoom???

**Comment by cmeng-git** [comment]
aTalk v 1.8.6 has been released with XEP-0048 Bookmarks implemented. Please refer to the online help on its implementation UI.

[[20190620] XEP-0048: Bookmarks and autoJoin option for conference](https://atalk.sytes.net/atalk/faq.html#feature_09)
